### DISCUSSION TITLE
[NextJS] Minor confusion about urqlState

### DISCUSSION THREAD

**Discussion by sysRoot** [urql-graphql/urql#2357] [Q&A] [answered]
```js
//package.json excerpt
    "next": "^12.1.0",
    "next-urql": "^3.3.2",
    "urql": "^2.2.0"
```

So, I have an App... it was wrapped `withUrqlClient`, because, well I assumed this was the expected behavior to get urqlState to be "picked up", however I'm left with a prop called urqlState and a subsection numbered object that stores my result as a string?

That's fine, that's what `JSON.parse()` is for, but this number seems to be pseudo randomly generated based on the queries structure. 

My whole progress was confounded by a change in the number when I restructured my query. Also, I see this as a problem in case something else causes the number to be changed. It also makes destructuring incredibly painful.

I also have index and multiple other pages, where I wanted to make getServerSideProps queries per page, consuming my database for only the page loaded. I'm trying to avoid State currently as there's no frontend method for editing/modifying the database. The frontend at this point merely consumes the API from Strapi.

The Docs for URQL SSR with Next implies I should not wrap the entire App with `withUrqlClient()`, specifically the section (that only mentions `getStaticProps()`) leaves me to infer that it should be wrapped around the lesser component, such as `index` (blog front page), `blog` (paginated list of articles), `poems` (same page as blog but consuming poems) from the api`, etc, e.g `export default withUrqlClient(config here)(Poems)` and setting `ssr: false` in the arguments passed into the HOC. 

This also causes me to infer, with little actual detail or information pertaining to it, that `urqlState` is then picked up by `withUrqlClient()` and does something seemingly automagical to adhere it to the props.

Below is the example from the docs with some minor additional comments to exemplify what I'm driving this question towards:
```js
import { withUrqlClient, initUrqlClient } from 'next-urql';
import { ssrExchange, dedupExchange, cacheExchange, fetchExchange, useQuery } from 'urql';


// Do I have to use useQuery? I am aiming to have props at rendering of the page. I had this working in apollo
const TODOS_QUERY = `
  query { todos { id text } }
`;

function Todos() {
  const [res] = useQuery({ query: TODOS_QUERY });
  return (
    <div>
      {res.data.todos.map(todo => (
        <div key={todo.id}>
          {todo.id} - {todo.text}
        </div>
      ))}
    </div>
  );
}

export async function getStaticProps(ctx) {
  const ssrCache = ssrExchange({ isClient: false });
  const client = initUrqlClient({
    url: 'your-url',
    exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange],
  });

  // This query is used to populate the cache for the query
  // used on this page.
  await client.query(TODOS_QUERY).toPromise();

  return {
    props: {
      // urqlState is a keyword here so withUrqlClient can pick it up.
      // what does this keyword do... All I have is empty pageProps and a urqlClient and resetUrqlClient prop
      // Wrapping the entire app with `withUrqlClient`, seems to make it work, however I'm left with a numbered cache result as a string? And as mentioned the docs imply it is better to do per page queries.
      urqlState: ssrCache.extractData(),
    },
    revalidate: 600,
  };
}

export default withUrqlClient(
  ssr => ({
    url: 'your-url',
  }),
  { ssr: false } // Important so we don't wrap our component in getInitialProps
)(Todos); // <<<<<< Lesser Component wrapped
```

Below is my implement of it, with my key redacted
```js
export async function getServerSideProps(ctx) {
  const ssrCache = ssrExchange({ isClient: false })
  const client = initUrqlClient({
    url: `https://a.meaningpoetry.com/graphql`,
    fetchOptions: {
      headers: {
        Authorization: `REDACTED-KEY`,
      },
    },
    exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange],
  })

  await client.query({ query: ARTICLES_QUERY, variables: { page: 1 } }).toPromise()
  return {
    props: {
      urqlState: ssrCache.extractData(),
    },
  }
}

export default withUrqlClient(
  (ssr) => ({
    url: 'https://a.meaningpoetry.com/graphql',
  }),
  { ssr: false } // Important so we don't wrap our component in getInitialProps
)(Home)
```

Is there something I'm missing, is there something automagical that should occur with `withUrqlClient(() => (), { ssr: false })(Component) ++ urqlState`. I suppose StateManagement may be necessary to some extent to mimic my Apollo setup which looked like the following: 

```js
export async function getServerSideProps() {
  const apolloClient = initializeApollo()

  const query = await apolloClient.query({
    query: POEMS_QUERY,
  })

  return addApolloState(apolloClient, {
    props: { poems: query.data.poems, pageCount: query.data.poemsConnection.aggregate.totalCount },
  })
}```

In Summary I'd really like to make use of the lighter urql, and nix dealing with the complexity of Apollo.
If I have to add in a State of sorts, I'll most likely make use of React's Context.
My main concern is the PRNG'd classifier on the cache, and why urqlState does not appear caught by withUrqlClient.

**Comment by JoviDeCroock** [comment, marked as answer]
You can do the same as you did with apollo by passing it down as `props: { poems: query.data.poems, pageCount: query.data.poemsConnection.aggregate.totalCount },` but that will mean that your client-side cache won't be hydrated. The fact that you can just use `useQuery` is because the cache is synchronously hydrated and subsequent page updates are faster for that reason.
Just do `const result = await client.query(TODOS_QUERY).toPromise();` and use it

So you can choose however you want to use it, i.e. just as a way to fetch things and not hydrate the cache. Using `useQuery` however will ensure that you always deal with the latest data in cache-updates, .... as `getServerSideProps` is a one-time thing rather than a continuous stream of information during client-actions

**Comment by sysRoot** [comment]
So what I am hearing for my use case scenario, is best to switch to useQuery, to keep information fresh. I can dig it. I don't want to be passing large swaths of state/context around, and would generally prefer fresh information per page. Plus it'd clean up my code a bit. And in this setup, I'd wrap each component I'd like to use useQuery in with the HOC? Thank you for a quick reply. That just leaves me having to rewrite my queries, and will make the overall refactor quick. I decided to migrate to StrapiV4 and stumbled upon urql while doing some searches and so far like it. didn't realize getServerSideProps was one-time like that. Thank you, apologize for user ignorance playing a large part in the issue here.

**Reply by JoviDeCroock** [reply]
No problem at all, that signal to me that we can improve on docs so that feedback is much appreciated. If you want `useQuery` to work wrap it with the HOC if you don't need it, just pass the data in to props
